pkg/resource/rule_groups_namespace/hooks.go

```diff
@@ -0,0 +1,342 @@
+	ErrRuleGroupsUpdating = errors.New("Rule Groups Namespace is in 'UPDATING' state, cannot be modified or deleted")
+)
+
+var (
+	// TerminalStatuses are the status strings that are a rule groups namespace
+	TerminalStatuses = []svcapitypes.RuleGroupsNamespaceStatusCode{
+		svcapitypes.RuleGroupsNamespaceStatusCode_CREATION_FAILED,
+		svcapitypes.RuleGroupsNamespaceStatusCode_UPDATE_FAILED,
+		svcapitypes.RuleGroupsNamespaceStatusCode_DELETING,
+	}
+)
+
+var (
+	requeueWaitWhileDeleting = ackrequeue.NeededAfter(
+		ErrRuleGroupsDeleting,
+		10*time.Second,
+	)
+	requeueWaitWhileCreating = ackrequeue.NeededAfter(
+		ErrRuleGroupsCreating,
+		15*time.Second,
+	)
+	requeueWaitWhileUpdating = ackrequeue.NeededAfter(
+		ErrRuleGroupsUpdating,
+		10*time.Second,
+	)
+)
+
+// ruleGroupsCreating returns true if the supplied rule groups is in the process
+// of being created
+func ruleGroupsCreating(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	ws := *r.ko.Status.Status.StatusCode
+	return ws == string(svcapitypes.RuleGroupsNamespaceStatusCode_CREATING)
+}
+
+// ruleGroupsCreating returns true if the supplied rule groups is in the process
+// of being deleted
+func ruleGroupsDeleting(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	ws := *r.ko.Status.Status.StatusCode
+	return ws == string(svcapitypes.RuleGroupsNamespaceStatusCode_DELETING)
+}
+
+// ruleGroupsCreating returns true if the supplied rule groups is in the process
+// of being updated
+func ruleGroupsUpdating(r *resource) bool {
+	if r.ko.Status.Status == nil {
+		return false
+	}
+	ws := *r.ko.Status.Status.StatusCode
+	return ws == string(svcapitypes.RuleGroupsNamespaceStatusCode_UPDATING)
+}
+
+// ruleGroupsHasTerminalStatus returns whether the supplied rule groups namespace is in a
+// terminal state
+func ruleGroupsHasTerminalStatus(r *resource) bool {
+	if r.ko.Status.Status.StatusCode == nil {
+		return false
+	}
+	ts := *r.ko.Status.Status.StatusCode
+	for _, s := range TerminalStatuses {
+		if ts == string(s) {
+			return true
+		}
+	}
+	return false
+}
+
+// customUpdateRuleGroups patches each of the resource properties in the backend AWS
+// service API and returns a new resource with updated fields.
+func (rm *resourceManager) customUpdateRuleGroups(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (*resource, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdateRuleGroups")
+	defer exit(err)
+
+	// Check if the state is being currently created, updated or deleted.
+	// If it is, then requeue because we can't update while it is in those states.
+	if ruleGroupsCreating(latest) {
+		return desired, requeueWaitWhileCreating
+	}
+	if ruleGroupsUpdating(latest) {
+		return desired, requeueWaitWhileUpdating
+	}
+	if ruleGroupsDeleting(latest) {
+		return desired, requeueWaitWhileDeleting
+	}
```

Comment:
tiny nit: Since a RuleGroupsNamespace can only have one of those statuses, how about using a switch/case here?